File: src/ui/viewer_3d.py
"""
基于 pyqtgraph.opengl.GLViewWidget 的自定义 3D 查看器核心组件。

负责相机管理、鼠标/键盘交互、缩放动画和视图矩阵重写。
网格/坐标轴由 GridRenderer 管理，点/轨迹由 TrackRenderer 管理。
"""

import logging
import os
from math import tan, radians

from PySide6.QtCore import Qt, QPoint, QTimer, QTime, Signal
from PySide6.QtGui import QVector3D, QMatrix4x4
import pyqtgraph.opengl as gl

from src.ui.grid_renderer import GridRenderer
from src.ui.track_renderer import TrackRenderer

logger = logging.getLogger(__name__)


class Viewer3D(gl.GLViewWidget):
    """
    自定义 3D 查看器组件。
    鼠标交互方式：
    - 左键拖动：旋转（轨道）
    - 右键拖动：平移
    - 中键点击：复位视角（动画过渡）
    - 滚轮：缩放
    """

    log_message = Signal(str)
    camera_changed = Signal()
    projection_mode_changed = Signal(bool)

    def __init__(self, parent=None):
        super().__init__(parent)

        # 相机初始状态
        self.initial_state = {
            'distance': 80,
            'elevation': 30,
            'azimuth': -135,
            'center': QVector3D(0, 0, 0),
            'pan_x': 0,
            'pan_y': -5,
            'scene_scale': 1.0,
        }

        self.setCameraPosition(
            distance=self.initial_state['distance'],
            elevation=self.initial_state['elevation'],
            azimuth=self.initial_state['azimuth'],
        )
        self.setBackgroundColor('#121212')

        self.scene_scale = self.initial_state['scene_scale']

        # 委托渲染器
        self.grid_renderer = GridRenderer(self)
        self.track_renderer = TrackRenderer(self)

        self.update_coordinate_system()

        self.mousePos = QPoint()

        self.pan_offset = QVector3D(
            self.initial_state['pan_x'],
            self.initial_state['pan_y'],
            0,
        )

        self.setFocusPolicy(Qt.StrongFocus)

        # 中键长按状态
        self.long_press_timer = QTimer(self)
        self.long_press_timer.setSingleShot(True)
        self.long_press_timer.timeout.connect(self.on_long_press_timeout)
        self.is_long_press = False

        self.animation_timer = QTimer(self)
        self.animation_timer.timeout.connect(self.update_animation)
        self.animation_start_time = None
        self.animation_duration = 800
        self.start_state = {}
        self.target_state = {}

        # 平滑缩放动画
        self._target_scene_scale = self.scene_scale
        self._zoom_anim_timer = QTimer(self)
        self._zoom_anim_timer.setInterval(16)
        self._zoom_anim_timer.timeout.connect(self._update_zoom_animation)

        # 投影模式（正交 / 透视）
        self._use_ortho = False
        self._ortho_blend = 0.0
        self._ortho_anim_timer = QTimer(self)
        self._ortho_anim_timer.setInterval(16)
        self._ortho_anim_timer.timeout.connect(self._update_ortho_animation)

        self.render_debug_enabled = (
            os.getenv("TRACER_RENDER_DEBUG", "0") == "1"
        )
        self.render_debug_verbose_point_updates = (
            os.getenv("TRACER_RENDER_DEBUG_VERBOSE", "0") == "1"
        )

    # ── Render debug ─────────────────────────────────────────────────

    def set_render_debug_options(self, enabled=None,
                                verbose_point_updates=None):
        if enabled is not None:
            self.render_debug_enabled = bool(enabled)
        if verbose_point_updates is not None:
            self.render_debug_verbose_point_updates = bool(
                verbose_point_updates,
            )

    # ── Delegation to GridRenderer ───────────────────────────────────

    def update_coordinate_system(self):
        dist = self.cameraParams()['distance']
        self.grid_renderer.update(dist, self.scene_scale)

    def _update_arrow_billboard(self):
        self.grid_renderer.update_arrow_billboard(
            scene_scale=self.scene_scale,
        )

    # ── Delegation to TrackRenderer ──────────────────────────────────

    def update_point(self, name, x, y, z, color=(1, 0, 0, 1), size=10):
        self.track_renderer.update_point(name, x, y, z, color, size)

    def clear_all(self):
        self.track_renderer.clear_all()

    def set_full_path_mode(self, enabled):
        self.track_renderer.set_full_path_mode(enabled)

    def set_trail_mode(self, enabled):
        self.track_renderer.set_trail_mode(enabled)

    def set_trail_length(self, length):
        self.track_renderer.set_trail_length(length)

    # ── Mouse events ─────────────────────────────────────────────────

    def mousePressEvent(self, ev):
        self.setFocus()
        self.mousePos = ev.pos()

        if ev.button() == Qt.MouseButton.MiddleButton:
            self.is_long_press = False
            self.long_press_timer.start(1000)
            ev.accept()
            return

        if (ev.button() == Qt.MouseButton.LeftButton
                or ev.button() == Qt.MouseButton.RightButton):
            if self.animation_timer.isActive():
                self.animation_timer.stop()
            ev.accept()
        else:
            super().mousePressEvent(ev)

    def mouseReleaseEvent(self, ev):
        if ev.button() == Qt.MouseButton.MiddleButton:
            if self.long_press_timer.isActive():
                self.long_press_timer.stop()
                if not self.is_long_press:
                    self.start_reset_animation(full_reset=False)
            self.is_long_press = False
            ev.accept()
        else:
            super().mouseReleaseEvent(ev)

    def on_long_press_timeout(self):
        self.is_long_press = True
        self.start_reset_animation(full_reset=True)

    def keyPressEvent(self, ev):
        if ev.key() == Qt.Key_R:
            self.auto_fit_view()
            ev.accept()
        else:
            super().keyPressEvent(ev)

    def mouseMoveEvent(self, ev):
        diff = ev.pos() - self.mousePos
        self.mousePos = ev.pos()

        if ev.buttons() == Qt.MouseButton.LeftButton:
            self.orbit(diff.x(), diff.y())
            self.update_coordinate_system()
            self.camera_changed.emit()

        elif ev.buttons() == Qt.MouseButton.RightButton:
            dist = self.cameraParams()['distance']
            scale = dist * 0.001
            self.pan_offset.setX(self.pan_offset.x() + diff.x() * scale)
            self.pan_offset.setY(self.pan_offset.y() - diff.y() * scale)
            self.update()

        else:
            super().mouseMoveEvent(ev)

    def wheelEvent(self, ev):
        if self.animation_timer.isActive():
            self.animation_timer.stop()

        delta = ev.angleDelta().y()
        factor = 1.1 if delta > 0 else (1.0 / 1.1)

        self._target_scene_scale *= factor
        if not self._zoom_anim_timer.isActive():
            self._zoom_anim_timer.start()
        ev.accept()

    # ── View matrix ──────────────────────────────────────────────────

    def viewMatrix(self):
        m = QMatrix4x4()
        m.translate(self.pan_offset.x(), self.pan_offset.y(), 0)
        m.translate(0, 0, -self.opts['distance'])
        m.rotate(self.opts['elevation'] - 90, 1, 0, 0)
        m.rotate(self.opts['azimuth'], 0, 0, 1)

        s = self.scene_scale
        m.scale(s, s, s)

        center = self.opts['center']
        m.translate(-center.x(), -center.y(), -center.z())

        return m

    # ── Projection matrix ─────────────────────────────────────────────

    def projectionMatrix(self, region, viewport):
        x0, y0, w, h = viewport
        dist = self.opts['distance']
        fov = self.opts['fov']
        nearClip = dist * 0.001
        farClip = dist * 1000.0

        r = nearClip * tan(0.5 * radians(fov))
        t = r * h / w

        left = r * ((region[0] - x0) * (2.0 / w) - 1)
        right = r * ((region[0] + region[2] - x0) * (2.0 / w) - 1)
        bottom = t * ((region[1] - y0) * (2.0 / h) - 1)
        top = t * ((region[1] + region[3] - y0) * (2.0 / h) - 1)

        if self._ortho_blend <= 0.0:
            tr = QMatrix4x4()
            tr.frustum(left, right, bottom, top, nearClip, farClip)
            return tr

        scale = dist / nearClip
        ol = left * scale
        or_ = right * scale
        ob = bottom * scale
        ot = top * scale

        if self._ortho_blend >= 1.0:
            tr = QMatrix4x4()
            tr.ortho(ol, or_, ob, ot, nearClip, farClip)
            return tr

        persp = QMatrix4x4()
        persp.frustum(left, right, bottom, top, nearClip, farClip)
        ortho = QMatrix4x4()
        ortho.ortho(ol, or_, ob, ot, nearClip, farClip)

        b = self._ortho_blend
        p = self._matrix4x4_to_list(persp)

        # 正交矩阵的 w 分量为常量 1，而透视矩阵的 w 分量与 z 相关。
        # 直接逐元素混合会在回切透视时生成数值尺度不一致的中间矩阵，
        # 导致动画帧出现裁剪/渲染异常。先做齐次等价缩放再混合可保持稳定。
        o = self._matrix4x4_to_list(ortho, scale=max(dist, 1e-6))

        blended = [p[i] + (o[i] - p[i]) * b for i in range(16)]
        return QMatrix4x4(*blended)

    @staticmethod
    def _matrix4x4_to_list(matrix, scale=1.0):
        values = []
        for row_idx in range(4):
            row = matrix.row(row_idx)
            values.extend([
                row.x() * scale,
                row.y() * scale,
                row.z() * scale,
                row.w() * scale,
            ])
        return values

    def toggle_projection(self):
        self._use_ortho = not self._use_ortho
        if not self._ortho_anim_timer.isActive():
            self._ortho_anim_timer.start()
        self.projection_mode_changed.emit(self._use_ortho)

    def _update_ortho_animation(self):
        target = 1.0 if self._use_ortho else 0.0
        lerp_speed = 0.18
        diff = target - self._ortho_blend
        if abs(diff) < 1e-4:
            self._ortho_blend = target
            self._ortho_anim_timer.stop()
        else:
            self._ortho_blend += diff * lerp_speed
        self.update_coordinate_system()
        self.camera_changed.emit()
        self.update()

    # ── Animation ────────────────────────────────────────────────────

    def _update_zoom_animation(self):
        lerp_speed = 0.28
        diff = self._target_scene_scale - self.scene_scale
        if abs(diff) / max(self.scene_scale, 1e-9) < 1e-4:
            self.scene_scale = self._target_scene_scale
            self._zoom_anim_timer.stop()
        else:
            self.scene_scale += diff * lerp_speed
        self.update_coordinate_system()
        self.update()

    def auto_fit_view(self):
        points = self.track_renderer.points
        if not points:
            return

        max_dist = 0
        for item in points.values():
            pos_data = item.pos
            if pos_data is None or len(pos_data) == 0:
                continue
            import numpy as np
            dists = np.linalg.norm(pos_data, axis=1)
            if len(dists) > 0:
                current_max = float(np.max(dists))
                if current_max > max_dist:
                    max_dist = current_max

        if max_dist < 1.0:
            target_dist = self.initial_state['distance']
        else:
            target_dist = max_dist * 3.5

        current_cam = self.cameraParams()
        self.start_state = {
            'distance': current_cam['distance'],
            'elevation': current_cam['elevation'],
            'azimuth': current_cam['azimuth'],
            'pan_x': self.pan_offset.x(),
            'pan_y': self.pan_offset.y(),
            'scene_scale': self.scene_scale,
        }
        self.target_state = {
            'distance': target_dist,
            'elevation': current_cam['elevation'],
            'azimuth': current_cam['azimuth'],
            'pan_x': self.initial_state['pan_x'],
            'pan_y': self.initial_state['pan_y'],
            'scene_scale': self.initial_state['scene_scale'],
        }

        self.animation_start_time = QTime.currentTime()
        self.animation_timer.start(16)

    def start_reset_animation(self, full_reset=True):
        if self._zoom_anim_timer.isActive():
            self._zoom_anim_timer.stop()

        current_params = self.cameraParams()

        if full_reset:
            target_distance = self.initial_state['distance']
            target_scene_scale = self.initial_state['scene_scale']
        else:
            target_distance = current_params['distance']
            target_scene_scale = self.scene_scale

        self._target_scene_scale = target_scene_scale

        current_azim = current_params['azimuth']
        target_azim = self.initial_state['azimuth']
        diff = target_azim - current_azim
        diff = (diff + 180) % 360 - 180
        effective_target_azim = current_azim + diff

        self.start_state = {
            'distance': current_params['distance'],
            'elevation': current_params['elevation'],
            'azimuth': current_params['azimuth'],
            'pan_x': self.pan_offset.x(),
            'pan_y': self.pan_offset.y(),
            'scene_scale': self.scene_scale,
        }
        self.target_state = {
            'distance': target_distance,
            'elevation': self.initial_state['elevation'],
            'azimuth': effective_target_azim,
            'pan_x': self.initial_state['pan_x'],
            'pan_y': self.initial_state['pan_y'],
            'scene_scale': target_scene_scale,
        }

        self.animation_start_time = QTime.currentTime()
        self.animation_timer.start(16)

    def animate_to_view(self, elevation, azimuth):
        """Smoothly animate the camera to the given orientation."""
        if self._zoom_anim_timer.isActive():
            self._zoom_anim_timer.stop()

        current_cam = self.cameraParams()
        current_azim = current_cam['azimuth']
        diff = azimuth - current_azim
        diff = (diff + 180) % 360 - 180
        effective_target_azim = current_azim + diff

        self.start_state = {
            'distance': current_cam['distance'],
            'elevation': current_cam['elevation'],
            'azimuth': current_cam['azimuth'],
            'pan_x': self.pan_offset.x(),
            'pan_y': self.pan_offset.y(),
            'scene_scale': self.scene_scale,
        }
        self.target_state = {
            'distance': current_cam['distance'],
            'elevation': elevation,
            'azimuth': effective_target_azim,
            'pan_x': self.pan_offset.x(),
            'pan_y': self.pan_offset.y(),
            'scene_scale': self.scene_scale,
        }

        self._target_scene_scale = self.scene_scale
        self.animation_start_time = QTime.currentTime()
        self.animation_timer.start(16)

    def update_animation(self):
        if not self.animation_start_time:
            return

        elapsed = self.animation_start_time.msecsTo(QTime.currentTime())
        progress = elapsed / self.animation_duration

        if progress >= 1.0:
            progress = 1.0
            self.animation_timer.stop()

        ease = 1 - (1 - progress) * (1 - progress)

        new_dist = (
            self.start_state['distance']
            + (self.target_state['distance'] - self.start_state['distance'])
            * ease
        )
        new_elev = (
            self.start_state['elevation']
            + (self.target_state['elevation'] - self.start_state['elevation'])
            * ease
        )
        new_azim = (
            self.start_state['azimuth']
            + (self.target_state['azimuth'] - self.start_state['azimuth'])
            * ease
        )

        new_pan_x = (
            self.start_state['pan_x']
            + (self.target_state['pan_x'] - self.start_state['pan_x']) * ease
        )
        new_pan_y = (
            self.start_state['pan_y']
            + (self.target_state['pan_y'] - self.start_state['pan_y']) * ease
        )
        self.pan_offset.setX(new_pan_x)
        self.pan_offset.setY(new_pan_y)

        start_scale = self.start_state.get('scene_scale', 1.0)
        target_scale = self.target_state.get('scene_scale', 1.0)
        self.scene_scale = start_scale + (target_scale - start_scale) * ease
        self._target_scene_scale = self.scene_scale

        self.setCameraPosition(
            distance=new_dist,
            elevation=new_elev,
            azimuth=new_azim,
        )
        self.update_coordinate_system()
        self.camera_changed.emit()
        self.update()

File: src/ui/opengl/Viewer3D.cpp
#include "Viewer3D.h"
#include "GridRenderer.h"
#include "TrackRenderer.h"
#include <QMouseEvent>
#include <QWheelEvent>
#include <QKeyEvent>
#include <QPainter>
#include <QDateTime>
#include <cmath>

// ============================================================
// Viewer3D.cpp — 3D OpenGL 视口实现
// ============================================================

Viewer3D::Viewer3D(QWidget* parent)
    : QOpenGLWidget(parent)
{
    // 启用深度测试格式
    QSurfaceFormat fmt;
    fmt.setVersion(3, 3);
    fmt.setProfile(QSurfaceFormat::CoreProfile);
    fmt.setDepthBufferSize(24);
    fmt.setSamples(4); // 4x MSAA 抗锯齿
    setFormat(fmt);

    setFocusPolicy(Qt::StrongFocus);

    // 缩放动画定时器（每 16ms ≈ 60fps）
    m_zoomAnimTimer.setInterval(16);
    connect(&m_zoomAnimTimer, &QTimer::timeout, this, &Viewer3D::updateZoomAnimation);

    // 正交/透视混合动画定时器
    m_orthoAnimTimer.setInterval(16);
    connect(&m_orthoAnimTimer, &QTimer::timeout, this, &Viewer3D::updateOrthoAnimation);

    // 相机动画定时器
    m_cameraAnimTimer.setInterval(16);
    connect(&m_cameraAnimTimer, &QTimer::timeout, this, &Viewer3D::updateCameraAnimation);

    // R 键长按定时器（1 秒恢复视角，2 秒恢复全部）
    m_resetViewTimer.setSingleShot(true);
    m_resetViewTimer.setInterval(1000);
    connect(&m_resetViewTimer, &QTimer::timeout, this, &Viewer3D::onResetViewTimeout);

    m_resetAllTimer.setSingleShot(true);
    m_resetAllTimer.setInterval(2000);
    connect(&m_resetAllTimer, &QTimer::timeout, this, &Viewer3D::onResetAllTimeout);
}

Viewer3D::~Viewer3D() {
    makeCurrent();
    delete m_gridRenderer;
    delete m_trackRenderer;
    doneCurrent();
}

// ── OpenGL 初始化 ─────────────────────────────────────────────

// 初始化 OpenGL 上下文和子渲染器
void Viewer3D::initializeGL() {
    initializeOpenGLFunctions();

    // 设置 3D 场景背景色（深黑色）
    glClearColor(0.071f, 0.071f, 0.071f, 1.0f);
    glEnable(GL_DEPTH_TEST);
    glEnable(GL_BLEND);
    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);

    // 初始化子渲染器（在 OpenGL 上下文内创建 VBO/VAO/Shader）
    m_gridRenderer  = new GridRenderer(this);
    m_trackRenderer = new TrackRenderer(this);
    m_gridRenderer->initialize();
    m_trackRenderer->initialize();

    m_gridRenderer->update(m_distance, m_sceneScale, m_elevation, m_azimuth, m_orthoBlend);
}

// 视口大小变化时更新参数
void Viewer3D::resizeGL(int w, int h) {
    m_viewportW = w;
    m_viewportH = h;
    glViewport(0, 0, w, h);
}

// 每帧渲染（OpenGL 内容 + QPainter overlay）
void Viewer3D::paintGL() {
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

    // 计算 MVP 矩阵
    QMatrix4x4 proj = computeProjectionMatrix();
    QMatrix4x4 view = computeViewMatrix();
    QMatrix4x4 mvp  = proj * view;

    // 渲染网格和坐标轴
    m_gridRenderer->render(mvp);

    // 渲染轨迹和点
    m_trackRenderer->render(mvp);

    // ── QPainter Overlay（文字标签）────────────────────────────
    // 注意：QPainter 必须在 OpenGL 绘制完成后创建，且需要调用
    // beginNativePainting/endNativePainting 保护 OpenGL 状态
    QPainter painter(this);
    painter.beginNativePainting();
    painter.endNativePainting();

    // 绘制坐标轴文字标签
    m_gridRenderer->renderLabels(&painter, mvp, m_viewportW, m_viewportH);
    painter.end();
}

// ── 矩阵计算 ──────────────────────────────────────────────────

QMatrix4x4 Viewer3D::computeViewMatrix() const {
    QMatrix4x4 m;
    m.translate(m_panX, m_panY, 0.0f);
    m.translate(0.0f, 0.0f, -float(m_distance));
    m.rotate(float(m_elevation - 90.0), 1.0f, 0.0f, 0.0f);
    m.rotate(float(m_azimuth), 0.0f, 0.0f, 1.0f);
    float s = m_sceneScale;
    m.scale(s, s, s);
    m.translate(-m_center);
    return m;
}

// 计算投影矩阵（支持透视/正交插值混合）
QMatrix4x4 Viewer3D::computeProjectionMatrix() const {
    float aspect = (m_viewportH > 0) ? float(m_viewportW) / m_viewportH : 1.0f;
    float nearClip = float(m_distance) * 0.001f;
    float farClip  = float(m_distance) * 1000.0f;
    if (nearClip < 0.001f) nearClip = 0.001f;

    if (m_orthoBlend <= 0.0f) {
        // 纯透视投影
        QMatrix4x4 p;
        p.perspective(m_fov, aspect, nearClip, farClip);
        return p;
    }
    if (m_orthoBlend >= 1.0f) {
        // 纯正交投影（等价缩放使场景大小与透视视图一致）
        float halfH = float(m_distance) * std::tan(m_fov * 0.5f * float(M_PI) / 180.0f);
        float halfW = halfH * aspect;
        QMatrix4x4 p;
        p.ortho(-halfW, halfW, -halfH, halfH, nearClip, farClip);
        return p;
    }

    // Homogeneous-scale blending (matching Python original):
    // The ortho matrix has constant w=1 while perspective has z-dependent w.
    // Naive element-wise lerp produces scale-inconsistent intermediate matrices.
    // Pre-scale the ortho matrix by dist to match the perspective matrix's
    // homogeneous magnitude, then blend.
    QMatrix4x4 perspM, orthoM;
    perspM.perspective(m_fov, aspect, nearClip, farClip);
    float halfH = float(m_distance) * std::tan(m_fov * 0.5f * float(M_PI) / 180.0f);
    float halfW = halfH * aspect;
    orthoM.ortho(-halfW, halfW, -halfH, halfH, nearClip, farClip);

    float b = m_orthoBlend;
    float dist = std::max(float(m_distance), 1e-6f);

    float p[16], o[16];
    for (int r = 0; r < 4; ++r)
        for (int c = 0; c < 4; ++c) {
            p[r * 4 + c] = perspM(r, c);
            o[r * 4 + c] = orthoM(r, c) * dist;
        }

    float blended[16];
    for (int i = 0; i < 16; ++i)
        blended[i] = p[i] + (o[i] - p[i]) * b;
    return QMatrix4x4(blended);
}

// ── 轨道相机旋转 ─────────────────────────────────────────────

// 处理鼠标左键拖动的轨道旋转
void Viewer3D::orbit(int dx, int dy) {
    m_azimuth   += dx * 0.5;
    m_elevation += dy * 0.5;
    // 仰角限制在 [-89, 89] 度，避免万向节死锁
    m_elevation = std::max(-89.0, std::min(89.0, m_elevation));
}

// ── 鼠标事件 ──────────────────────────────────────────────────

void Viewer3D::mousePressEvent(QMouseEvent* ev) {
    setFocus();
    m_lastMousePos = ev->pos();

    if (ev->button() == Qt::LeftButton || ev->button() == Qt::RightButton) {
        if (m_cameraAnimTimer.isActive()) m_cameraAnimTimer.stop();
        ev->accept();
    }
}

void Viewer3D::mouseReleaseEvent(QMouseEvent* ev) {
    Q_UNUSED(ev);
}

void Viewer3D::mouseMoveEvent(QMouseEvent* ev) {
    QPoint curr = ev->pos();
    int dx = curr.x() - m_lastMousePos.x();
    int dy = curr.y() - m_lastMousePos.y();
    m_lastMousePos = curr;

    if (ev->buttons() & Qt::LeftButton) {
        orbit(dx, dy);
        m_gridRenderer->update(m_distance, m_sceneScale, m_elevation, m_azimuth, m_orthoBlend);
        emit cameraChanged();
        update();
    } else if (ev->buttons() & Qt::RightButton) {
        float scale = float(m_distance) * 0.001f;
        m_panX += dx * scale;
        m_panY -= dy * scale;
        update();
    }
}

void Viewer3D::wheelEvent(QWheelEvent* ev) {
    if (m_cameraAnimTimer.isActive()) m_cameraAnimTimer.stop();

    int delta = ev->angleDelta().y();
    float factor = (delta > 0) ? 1.1f : (1.0f / 1.1f);
    m_targetSceneScale *= factor;
    if (!m_zoomAnimTimer.isActive()) m_zoomAnimTimer.start();
    ev->accept();
}

void Viewer3D::keyPressEvent(QKeyEvent* ev) {
    if (ev->key() == Qt::Key_R && !ev->isAutoRepeat()) {
        m_rKeyHeld       = true;
        m_resetViewFired = false;
        m_resetViewTimer.start();
        m_resetAllTimer.start();
        ev->accept();
    } else {
        QOpenGLWidget::keyPressEvent(ev);
    }
}

void Viewer3D::keyReleaseEvent(QKeyEvent* ev) {
    if (ev->key() == Qt::Key_R && !ev->isAutoRepeat()) {
        m_rKeyHeld = false;
        m_resetViewTimer.stop();
        m_resetAllTimer.stop();
        ev->accept();
    } else {
        QOpenGLWidget::keyReleaseEvent(ev);
    }
}

void Viewer3D::onResetViewTimeout() {
    if (!m_rKeyHeld) return;
    m_resetViewFired = true;
    startResetAnimation(false);
}

void Viewer3D::onResetAllTimeout() {
    if (!m_rKeyHeld) return;
    startResetAnimation(true);
}

// ── 动画 ──────────────────────────────────────────────────────

void Viewer3D::updateZoomAnimation() {
    float diff = m_targetSceneScale - m_sceneScale;
    float relDiff = std::abs(diff) / std::max(m_sceneScale, 1e-9f);
    if (relDiff < 1e-4f) {
        m_sceneScale = m_targetSceneScale;
        m_zoomAnimTimer.stop();
    } else {
        m_sceneScale += diff * 0.28f;
    }
    m_gridRenderer->update(m_distance, m_sceneScale, m_elevation, m_azimuth, m_orthoBlend);
    update();
}

// 正交/透视混合动画（指数插值）
void Viewer3D::updateOrthoAnimation() {
    float target = m_useOrtho ? 1.0f : 0.0f;
    float diff   = target - m_orthoBlend;
    if (std::abs(diff) < 1e-4f) {
        m_orthoBlend = target;
        m_orthoAnimTimer.stop();
    } else {
        m_orthoBlend += diff * 0.18f;
    }
    m_gridRenderer->update(m_distance, m_sceneScale, m_elevation, m_azimuth, m_orthoBlend);
    emit cameraChanged();
    update();
}

void Viewer3D::updateCameraAnimation() {
    qint64 now     = QDateTime::currentMSecsSinceEpoch();
    qint64 elapsed = now - m_animStartTime;
    float  t       = float(elapsed) / m_animDuration;
    if (t >= 1.0f) {
        t = 1.0f;
        m_cameraAnimTimer.stop();
    }
    float ease = 1.0f - (1.0f - t) * (1.0f - t);

    auto lerp = [](double a, double b, float e) { return a + (b - a) * e; };
    auto lerpf= [](float  a, float  b, float e) { return a + (b - a) * e; };

    m_distance   = lerp(m_animStart.distance,   m_animTarget.distance,   ease);
    m_elevation  = lerp(m_animStart.elevation,  m_animTarget.elevation,  ease);
    m_azimuth    = lerp(m_animStart.azimuth,    m_animTarget.azimuth,    ease);
    m_panX       = lerpf(float(m_animStart.panX), float(m_animTarget.panX), ease);
    m_panY       = lerpf(float(m_animStart.panY), float(m_animTarget.panY), ease);
    m_sceneScale = lerpf(float(m_animStart.sceneScale), float(m_animTarget.sceneScale), ease);
    m_targetSceneScale = m_sceneScale;

    m_gridRenderer->update(m_distance, m_sceneScale, m_elevation, m_azimuth, m_orthoBlend);
    emit cameraChanged();
    update();
}

// 切换透视/正交投影（触发混合动画）
void Viewer3D::toggleProjection() {
    m_useOrtho = !m_useOrtho;
    if (!m_orthoAnimTimer.isActive()) m_orthoAnimTimer.start();
    emit projectionModeChanged(m_useOrtho);
}

void Viewer3D::autoFitView() {
    const auto& pts = m_trackRenderer->points();
    if (pts.isEmpty()) return;

    float maxDist = 0.0f;
    for (auto it = pts.cbegin(); it != pts.cend(); ++it) {
        for (const QVector3D& p : it.value().history) {
            maxDist = std::max(maxDist, p.length());
        }
    }

    double targetDist = (maxDist < 1.0f) ? INIT_DISTANCE : maxDist * 3.5;

    m_animStart  = {m_distance, m_elevation, m_azimuth, m_panX, m_panY, m_sceneScale};
    m_animTarget = {targetDist, m_elevation, m_azimuth, INIT_PAN_X, INIT_PAN_Y, INIT_SCENE_SCALE};
    m_animStartTime = QDateTime::currentMSecsSinceEpoch();
    if (!m_cameraAnimTimer.isActive()) m_cameraAnimTimer.start();
}

void Viewer3D::startResetAnimation(bool fullReset) {
    if (m_zoomAnimTimer.isActive()) m_zoomAnimTimer.stop();

    double targetDist  = fullReset ? INIT_DISTANCE : m_distance;
    float  targetScale = fullReset ? INIT_SCENE_SCALE : m_sceneScale;
    m_targetSceneScale = targetScale;

    double diffAzim = INIT_AZIMUTH - m_azimuth;
    diffAzim = std::fmod(diffAzim + 180.0, 360.0) - 180.0;

    m_animStart  = {m_distance, m_elevation, m_azimuth, m_panX, m_panY, m_sceneScale};
    m_animTarget = {targetDist, INIT_ELEVATION, m_azimuth + diffAzim,
                    INIT_PAN_X, INIT_PAN_Y, targetScale};
    m_animStartTime = QDateTime::currentMSecsSinceEpoch();
    if (!m_cameraAnimTimer.isActive()) m_cameraAnimTimer.start();
}

void Viewer3D::animateToView(double elevation, double azimuth) {
    if (m_zoomAnimTimer.isActive()) m_zoomAnimTimer.stop();

    double diffAzim = azimuth - m_azimuth;
    diffAzim = std::fmod(diffAzim + 180.0, 360.0) - 180.0;

    m_animStart  = {m_distance, m_elevation, m_azimuth, m_panX, m_panY, m_sceneScale};
    m_animTarget = {m_distance, elevation, m_azimuth + diffAzim, m_panX, m_panY, m_sceneScale};
    m_targetSceneScale = m_sceneScale;
    m_animStartTime = QDateTime::currentMSecsSinceEpoch();
    if (!m_cameraAnimTimer.isActive()) m_cameraAnimTimer.start();
}

// ── 公有接口委托 ──────────────────────────────────────────────

void Viewer3D::updatePoint(const QString& name, double x, double y, double z,
                             const QColor& color, int size) {
    if (m_trackRenderer) m_trackRenderer->updatePoint(name, x, y, z, color, size);
    update();
}

void Viewer3D::clearAll() {
    if (m_trackRenderer) m_trackRenderer->clearAll();
    update();
}

void Viewer3D::setFullPathMode(bool e) { if (m_trackRenderer) m_trackRenderer->setFullPathMode(e); update(); }
void Viewer3D::setTrailMode(bool e)    { if (m_trackRenderer) m_trackRenderer->setTrailMode(e);    update(); }
void Viewer3D::setTrailLength(int l)   { if (m_trackRenderer) m_trackRenderer->setTrailLength(l);  update(); }

void Viewer3D::setRenderDebugOptions(bool enabled, bool verbosePointUpdates) {
    m_renderDebugEnabled        = enabled;
    m_renderDebugVerboseUpdates = verbosePointUpdates;
}
